quinn-proto/src/connection/streams.rs

```diff
@@ -236,6 +236,34 @@ impl Streams {
         }
     }
 
+    pub fn read_chunk(&mut self, id: StreamId) -> Result<Option<ReadChunkResult>, ReadError> {
+        let mut entry = match self.recv.entry(id) {
+            hash_map::Entry::Vacant(_) => return Err(ReadError::UnknownStream),
+            hash_map::Entry::Occupied(e) => e,
+        };
+        let rs = entry.get_mut();
+        match rs.read_chunk() {
```

Comment:
This is pretty repetitive, can we factor out a bunch of the match here (ideally in a separate commit)? Could apply to `read_unordered()` as well, ideally, but just between `read()` and `read_chunk()` would be good if that makes it hard.

Reply:
I got a solution I find a little complex in feb3b63, what do you think about this?

Reply:
It looks like you can just inline the `IntoReadResult` into the closure passed to `try_read()`? We already have the nested `map()` calls anyway. Or am I missing something?

Reply:
Ho yeah maybe ! :sweat_smile:
For the moment I get stuck because `self.add_read_credits(len as u64)` needs to mutate `self` to compute `transmit_max_data`, and it is already mutably borrowed by the closure. But I'll try a little harder in a bit.